File: Code/36.有效的数独.cpp
/*
 * @lc app=leetcode.cn id=36 lang=cpp
 *
 * [36] 有效的数独
 */
#include <iostream>
#include <unordered_set>
using namespace std;
// @lc code=start
class Solution {
 public:
  bool isValidSudoku(vector<vector<char>>& board) {
    unordered_set<char> row[9];
    unordered_set<char> col[9];
    unordered_set<char> box[9];
    for (int i = 0; i < 9; i++) {
      for (int j = 0; j < 9; j++) {
        char digit = board[i][j];
        if (digit == '.') continue;
        if (row[i].find(digit) != row[i].end()) return false;
        row[i].insert(digit);
        if (col[j].find(digit) != col[j].end()) return false;
        col[j].insert(digit);

        int k = (i / 3) * 3 + j / 3;  // 将二维转为一维数组
        if (box[k].find(digit) != box[k].end()) return false;
        box[k].insert(digit);
      }
    }
    return true;
  }
};
// @lc code=end

File: Code/36.有效的数独.java
/*
 * @lc app=leetcode.cn id=36 lang=java
 *
 * [36] 有效的数独
 */

// @lc code=start
class Solution {
  public boolean isValidSudoku(char[][] board) {
    int[][] row = new int[9][9];
    int[][] col = new int[9][9];
    int[][][] box = new int[3][3][9];
    for (int i = 0; i < 9; i++) {
      for (int j = 0; j < 9; j++) {
        char c = board[i][j];
        if (c != '.') {
          int index = c - '0' - 1;
          row[i][index]++;
          col[j][index]++;
          box[i / 3][j / 3][index]++;
          if (row[i][index] > 1 || col[j][index] > 1 || box[i / 3][j / 3][index] > 1) {
            return false;
          }
        }
      }
    }
    return true;

  }
}
// @lc code=end
